zone/parcels.cpp

```diff
@@ -0,0 +1,755 @@
+
+			if (send_to_client.at(0).character_name.empty()) {
+				MessageString(
+					Chat::Yellow,
+					PARCEL_UNKNOWN_NAME,
+					merchant->GetCleanName(),
+					parcel_in->send_to,
+					money.c_str()
+				);
+				DoParcelCancel();
+				SendParcelAck();
+				return;
+			}
+
+			ParcelsRepository::Parcels parcel_out;
+			parcel_out.from_name = GetName();
+			parcel_out.note      = parcel_in->note;
+			parcel_out.sent_date = time(nullptr);
+			parcel_out.quantity  = parcel_in->quantity;
+			parcel_out.item_id   = PARCEL_MONEY_ITEM_ID;
+			parcel_out.to_name   = parcel_in->send_to;
+			parcel_out.slot_id   = next_slot;
+			parcel_out.id        = 0;
+
+			auto result = ParcelsRepository::InsertOne(database, parcel_out);
+			if (!result.id) {
+				LogError(
+					"Failed to add parcel to database.  From {} to {} item {} quantity {}",
+					parcel_out.from_name,
+					parcel_out.to_name,
+					parcel_out.item_id,
+					parcel_out.quantity
+				);
+				Message(
+					Chat::Yellow,
+					"Unable to save parcel to the database. Please see an administrator."
+				);
+				return;
+			}
+
+			MessageString(
+				Chat::Yellow,
+				PARCEL_DELIVERY,
+				merchant->GetCleanName(),
+				money.c_str(),
+				send_to_client.at(0).character_name.c_str()
+			);
+
+			if (player_event_logs.IsEventEnabled(PlayerEvent::PARCEL_SEND)) {
+				PlayerEvent::ParcelSend e{};
+				e.from_player_name = parcel_out.from_name;
+				e.to_player_name   = parcel_out.to_name;
+				e.item_id          = parcel_out.item_id;
+				e.quantity         = parcel_out.quantity;
+				e.sent_date        = parcel_out.sent_date;
+
+				RecordPlayerEventLog(PlayerEvent::PARCEL_SEND, e);
+			}
+
+			m_parcel_platinum = 0;
+			m_parcel_gold     = 0;
+			m_parcel_silver   = 0;
+			m_parcel_copper   = 0;
+			auto outapp = new EQApplicationPacket(OP_FinishTrade);
+			FastQueuePacket(&outapp);
+
+			Parcel_Struct ps{};
+			ps.item_slot = parcel_out.slot_id;
+			strn0cpy(ps.send_to, parcel_out.to_name.c_str(), sizeof(ps.send_to));
+
+			SendParcelDeliveryToWorld(ps);
+
+			break;
+		}
+	}
+
+	SendParcelAck();
+	SendParcelIconStatus();
+	SetParcelEnabled(false);
+	GetParcelTimer()->Enable();
+}
+
+void Client::SendParcelAck()
+{
+	auto outapp = new EQApplicationPacket(OP_FinishTrade);
+	FastQueuePacket(&outapp);
+
+	outapp = new EQApplicationPacket(OP_ShopSendParcel, sizeof(Parcel_Struct));
+	auto data = (Parcel_Struct *) outapp->pBuffer;
+	data->item_slot = 0xffffffff;
+	data->quantity  = 0xffffffff;
+	FastQueuePacket(&outapp);
+}
+
+void Client::SendParcelRetrieveAck()
+{
+	auto outapp = new EQApplicationPacket(OP_ShopRetrieveParcel);
+	FastQueuePacket(&outapp);
+}
+
+void Client::SendParcelDeliveryToWorld(Parcel_Struct parcel)
+{
+	auto out  = new ServerPacket(ServerOP_ParcelDelivery, sizeof(Parcel_Struct));
+	auto data = (Parcel_Struct *) out->pBuffer;
+
+	data->item_slot = parcel.item_slot;
+	strn0cpy(data->send_to, parcel.send_to, sizeof(data->send_to));
+
+	worldserver.SendPacket(out);
+}
```

Comment:
Memory leak on `out` here

Reply:
As noted above, converted to unique_ptr.